### PR TITLE
storage: ensure we fetch a recent upper on startup

### PR DESCRIPTION
**Pull request by guswynn**
Attempt at dealing with https://github.com/MaterializeInc/materialize/issues/22725

We believe this to be the cause of https://materializeinc.slack.com/archives/C01CFKM1QRF/p1698391779182889,
as write handles can be arbitrarily delayed.

### Motivation

  * This PR fixes a recognized bug.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration or tests, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

File: src/storage-controller/src/lib.rs [hunk 1]
```diff
@@ -2052,7 +2052,7 @@ where
             shard_name: id.to_string(),
             handle_purpose: format!("controller data for {}", id),
         };
-        let write = persist_client
+        let mut write = persist_client
```
Comment (bkirwi) [inline src/storage-controller/src/lib.rs:2055]: It it okay to move the write-handle open to _after_ the read handle open? It sounds like it's important for the write handle to start out newer than the read handle, semantically, and that would help make that more obvious. (And probably be enough to fix the bug all on its own, though I think the explicit call is not a bad idea...)

Comment (guswynn) [reply src/storage-controller/src/lib.rs:2055]: yeah, let me go ahead and make this change

Comment (guswynn) [reply src/storage-controller/src/lib.rs]: done!

**Review by bkirwi — APPROVED** [review]
🆒

**Pull request merged**